src/libcore/ptr/mod.rs

```diff
@@ -795,6 +801,7 @@ pub unsafe fn read_unaligned<T>(src: *const T) -> T {
 #[inline]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub unsafe fn write<T>(dst: *mut T, src: T) {
+    debug_assert!(is_aligned_and_not_null(dst), "attempt to write to unaligned or null pointer");
```

Comment:
I wonder if this is the problem, maybe try landing `swap_nonoverlapping` and this separately.

Maybe some uses of `ptr::write`  want to use `intrinsics::move_val_init` directly without creating a raw pointer? This is so important, IIRC we lower `intrinsics::move_val_init` in MIR building directly.